index.php

```diff
@@ -158,22 +158,115 @@ function gutenberg_register_scripts() {
 
 	// Editor Scripts.
-	wp_register_script( 'tinymce-nightly', 'https://fiddle.azurewebsites.net/tinymce/nightly/tinymce' . $suffix . '.js' );
-	wp_register_script( 'wp-i18n', plugins_url( 'i18n/build/index.js', __FILE__ ), array(), filemtime( plugin_dir_path( __FILE__ ) . 'i18n/build/index.js' ) );
-	wp_register_script( 'wp-element', plugins_url( 'element/build/index.js', __FILE__ ), array( 'react', 'react-dom', 'react-dom-server' ), filemtime( plugin_dir_path( __FILE__ ) . 'element/build/index.js' ) );
-	wp_register_script( 'wp-blocks', plugins_url( 'blocks/build/index.js', __FILE__ ), array( 'wp-element', 'tinymce-nightly' ), filemtime( plugin_dir_path( __FILE__ ) . 'blocks/build/index.js' ) );
+	gutenberg_register_vendor_script(
+		'tinymce-nightly',
+		'https://fiddle.azurewebsites.net/tinymce/nightly/tinymce' . $suffix . '.js'
+	);
+	wp_register_script(
+		'wp-i18n',
+		plugins_url( 'i18n/build/index.js', __FILE__ ),
+		array(),
+		filemtime( plugin_dir_path( __FILE__ ) . 'i18n/build/index.js' )
+	);
+	wp_register_script(
+		'wp-element',
+		plugins_url( 'element/build/index.js', __FILE__ ),
+		array( 'react', 'react-dom', 'react-dom-server' ),
+		filemtime( plugin_dir_path( __FILE__ ) . 'element/build/index.js' )
+	);
+	wp_register_script(
+		'wp-blocks',
+		plugins_url( 'blocks/build/index.js', __FILE__ ),
+		array( 'wp-element', 'tinymce-nightly' ),
+		filemtime( plugin_dir_path( __FILE__ ) . 'blocks/build/index.js' )
+	);
 
 	// Editor Styles.
-	wp_register_style( 'wp-blocks', plugins_url( 'blocks/build/style.css', __FILE__ ), array(), filemtime( plugin_dir_path( __FILE__ ) . 'blocks/build/style.css' ) );
+	wp_register_style(
+		'wp-blocks',
+		plugins_url( 'blocks/build/style.css', __FILE__ ),
+		array(),
+		filemtime( plugin_dir_path( __FILE__ ) . 'blocks/build/style.css' )
+	);
 }
 add_action( 'init', 'gutenberg_register_scripts' );
 
 /**
+ * Registers a vendor script from a URL, preferring a locally cached version if
+ * possible, or downloading it if the cached version is unavailable or
+ * outdated.
+ *
+ * @param  string $handle Name of the script.
+ * @param  string $src    Full URL of the external script.
+ * @param  array  $deps   Optional. An array of registered script handles this
+ *                        script depends on.
+ *
+ * @since 0.1.0
+ */
+function gutenberg_register_vendor_script( $handle, $src, $deps = array() ) {
+	$filename = basename( $src );
+	$full_path = plugin_dir_path( __FILE__ ) . 'vendor/' . $filename;
+
+	$needs_fetch = (
+		! file_exists( $full_path ) ||
+		time() - filemtime( $full_path ) >= DAY_IN_SECONDS
+	);
+
+	if ( $needs_fetch ) {
+		// Determine whether we can write to this file.  If not, don't waste
+		// time doing a network request.
+		$f = fopen( $full_path, 'a' );
+		if ( ! $f ) {
+			// Failed to open the file for writing, probably due to server
+			// permissions.  Enqueue the script directly from the URL instead.
+			// Note: the `fopen` call above will have generated a warning.
```

Comment:
I am not sure how long this code will stick around.  I don't have a strong preference either way, but if this is primarily in place during early development, this warning seemed potentially helpful for power users.